modules/3d/include/opencv2/3d/detail/submap.hpp

```diff
@@ -0,0 +1,544 @@
+// This file is part of OpenCV project.
+// It is subject to the license terms in the LICENSE file found in the top-level directory
+// of this distribution and at http://opencv.org/license.html
+
+#ifndef __OPENCV_RGBD_SUBMAP_HPP__
+#define __OPENCV_RGBD_SUBMAP_HPP__
+
+#include <opencv2/core/cvdef.h>
+
+#include <opencv2/core/affine.hpp>
+#include <type_traits>
+#include <vector>
+#include <map>
+#include <unordered_map>
+
+#include "opencv2/core/mat.inl.hpp"
+#include "opencv2/3d/detail/pose_graph.hpp"
+
+namespace cv
+{
+namespace detail
+{
+template<typename MatType>
+class Submap
+{
+   public:
+    struct PoseConstraint
+    {
+        Affine3f estimatedPose;
+        int weight;
+
+        PoseConstraint() : weight(0){};
+
+        void accumulatePose(const Affine3f& _pose, int _weight = 1)
+        {
+            Matx44f accPose = estimatedPose.matrix * weight + _pose.matrix * _weight;
+            weight         += _weight;
+            accPose        /= float(weight);
+            estimatedPose   = Affine3f(accPose);
+        }
+    };
+    typedef std::map<int, PoseConstraint> Constraints;
+
+    Submap(int _id, const VolumeParams& volumeParams, const cv::Affine3f& _pose = cv::Affine3f::Identity(),
+           int _startFrameId = 0)
+        : id(_id), pose(_pose), cameraPose(Affine3f::Identity()), startFrameId(_startFrameId)
+    {
+        VolumeParams vp = volumeParams;
+        vp.kind = VolumeParams::VolumeKind::HASHTSDF;
+        Ptr<VolumeParams> pvp = makePtr<VolumeParams>(vp);
+        volume = makeVolume(pvp);
+
+        CV_LOG_INFO(NULL, "Created volume");
```

Comment:
Logging should not be used for tracing tasks.

Reply:
removed